docs/dialect/ops_reference.rst

```diff
@@ -31,45 +32,1304 @@ The `CSR64` encoding is usually defined with the alias::
+
+Example:
+^^^^^^^^
+
+.. code-block:: text
+
+    %thunk = constant 0.0 : f64
+    %answer = graphblas.apply_generic %sparse_tensor : tensor<?xf64, #CV64> to tensor<?xf64, #CV64> {
+      ^bb0(%val: f64):
+        %pick = cmpf olt, %val, %thunk : f64
+        %result = select %pick, %val, %thunk : f64
+        graphblas.yield transform_out %result : f64
+    }
+
+Syntax:
+^^^^^^^
+
+.. code-block:: text
+
+    operation ::= `graphblas.apply_generic` $input attr-dict `:` type($input) `to` type($output) $extensions
+
+
+Operands:
+^^^^^^^^^
+
+.. list-table::
+   :widths: 25 75
+   :header-rows: 1
+
+   * - Operand
+     - Description
+   * - ``input``
+     - Input tensor (CSR matrix, CSC matrix, or sparse vector)
+   * - ``extensions``
+     - Variadic list of regions describing the transformation behavior. Must have size 1.
+
+
+Results:
+^^^^^^^^
+
+.. list-table::
+   :widths: 25 75
+   :header-rows: 1
+
+   * - Result
+     - Description
+   * - ``output``
+     - Output tensor
+
+
+``graphblas.matrix_multiply``
+-----------------------------
+
+Performs a matrix multiply according to the given semiring and optional structural mask.
+The structural mask specifies which values in the output are to be computed and thus must
+have the same shape as the expected output.
+
+The semiring must be a string of the form "<ADD_NAME>_<MUL_NAME>", e.g. "plus_times".
+The options for "<ADD_NAME>" are "plus", "any", and "min". The options for
+"<MUL_NAME>" are "pair", "times", "plus", "first", and "second". 
+
+If the first input is a matrix, it must be CSR format. If the second input
+is a matrix, it must be CSC format.  Matrix times vector will return a vector.
+Vector times matrix will return a vector.  Matrix times matrix will return
+a CSR matrix.
+
+The mask (if provided) must be the same format as the returned object. There's an optional
+boolean ``mask_complement`` attribute (which has a default value of ``false``) that will make
+the op use the complement of the mask.
+	
+Example:
+^^^^^^^^
+
+.. code-block:: text
+
+    %answer = graphblas.matrix_multiply %argA, %argB { semiring = "plus_plus" } : (tensor<?x?xi64, #CSR64>, tensor<?x?xi64, #CSC64>) to tensor<?x?xi64, #CSR64>
+
+.. code-block:: text
+
+    %answer = graphblas.matrix_multiply %argA, %argB, %mask { semiring = "min_times" } : (tensor<?x?xf64, #CSR64>, tensor<?x?xf64, #CSC64>, tensor<?x?xf64, #CSR64>) to tensor<?x?xf64, #CSR64>
+
+.. code-block:: text
+
+    %answer = graphblas.matrix_multiply %mat, %vec, %mask { semiring = "any_first", mask_complement = true } : (tensor<?x?xf64, #CSR64>, tensor<?xf64, #CV64>, tensor<?x?xf64, #CSR64>) to tensor<?x?xf64, #CSR64>
+
+.. code-block:: text
+
+    %answer = graphblas.matrix_multiply %vec, %mat { semiring = "min_second", mask_complement = true } : (tensor<?xf64, #CV64>, tensor<?x?xf64, #CSR64>) to tensor<?x?xf64, #CSR64>
```

Comment:
Output needs to be a vector. The matrix argument should also be CSC because of its position.

Reply:
https://github.com/metagraph-dev/mlir-graphblas/commit/9060908223f773aaab02742d549299dde029570f implements this suggestion.